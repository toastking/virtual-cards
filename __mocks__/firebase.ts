import { mockFirebase } from 'ts-mock-firebase';

const firebase = mockFirebase();

export default firebase;
