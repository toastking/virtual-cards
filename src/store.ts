import Vue from 'vue';
import Vuex from 'vuex';
import { vuexfireMutations } from 'vuexfire';
import { DeckModule, DeckState } from './store/modules/deck';
import { GameModule, GameState, LoadingStatus } from './store/modules/game';
import { HistoryModule, HistoryState } from './store/modules/history';
import { Player, PlayerModule, PlayerState } from './store/modules/player';

Vue.use(Vuex);

export interface State {
  game: GameState;
  player: PlayerState;
  deck: DeckState;
  history: HistoryState;
}

export default new Vuex.Store<State>({
  mutations: {
    ...vuexfireMutations,
  },
  actions: {
    async doTurn({ dispatch, commit, getters }) {
      commit('turnRequestStarted');

      // Draw a card, wait a little bit, then go to the next player
      await dispatch('drawCard');
      await dispatch('nextPlayer');

      commit('turnRequestDone');
    },
  },
  getters: {
    decks(state) {
      return state.deck.decks;
    },
    gameOver(state) {
      return state.game.game.gameCompleted;
    },
    isYourTurn(state): boolean {
      return (
        state.game.game.currentPlayer === state.player.userPlayerId &&
        !state.game.game.gameCompleted
      );
    },
    turnIsLoading(state): boolean {
      return state.game.turnLoadingState === LoadingStatus.LOADING;
    },
    currentPlayer(state): Player | undefined {
      const id = state.game.game.currentPlayer;
      return state.player.players.find(player => id === player.id);
    },
    historyWithPlayerInfo(state) {
      // map the player info to an id
      const playerMapping: Map<string, Player> = new Map();

      for (const player of state.player.players) {
        const { id } = player;
        if (id) {
          playerMapping.set(id, player);
        }
      }

      if (state.history.history) {
        return state.history.history.map(entry => ({
          ...entry,
          player: playerMapping.get(entry.playerId),
        }));
      }
      return [];
    },
  },
  modules: {
    game: GameModule,
    player: PlayerModule,
    deck: DeckModule,
    history: HistoryModule,
  },
});
