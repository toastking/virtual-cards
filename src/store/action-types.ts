/** Start the create game flow */
export const createGame = 'createGame';
/** Setup the firestore bindings for a game */
export const setupGameBinding = 'setupGameBinding';
