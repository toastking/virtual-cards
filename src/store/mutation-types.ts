/** Game creation request succeeds */
export const CREATE_GAME_SUCCESS = 'Create Game Success';
/** When the create game request fails */
export const CREATE_GAME_FAILURE = 'Create Game Failure';
