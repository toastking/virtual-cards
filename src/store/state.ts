import { firestore } from 'firebase';

/** A Player's icon */
export enum Avatar {
  NONE,
  ALPACA,
  BUFALLO,
  GIRAFFE,
  OTTER,
  SHEEP,
}

/** Player in the game */
export interface Player {
  name: string;
  avatar: Avatar;
  id?: string;
}

/** State for current players in a game */
export interface PlayerState {
  players: Player[];
  /** The player id for the current user */
  userPlayerId: string;
}

/** Represents loading state so we can show loading spinners */
export enum LoadingStatus {
  NOT_STARTED,
  LOADING,
  OK,
  ERROR,
}

/** Represents the general state of a card game */
export interface Game {
  currentPlayer: string | null;
  gameCompleted: boolean;
  gameStarted: boolean;
}

export interface GameState {
  gameId: string | null;
  gameLoadingStatus: LoadingStatus;
  game: Game;
  // Loading for the request for the turn
  turnLoadingState: LoadingStatus;
}

/** A record of what card was drawn by what user */
export interface HistoryEntry {
  playerId: string;
  /** String representation of a card based on vue-plaing-card */
  card: string;
  timestamp: firestore.Timestamp;
}

/** State for which card was drawn and when */
export interface HistoryState {
  history: HistoryEntry[] | null;
}

/** Reprents a pile of cards, for something like kings */
export interface Deck {
  drawnCards: string[];
  currentCard?: string;
  id?: string; // Id from firestore
}

/** Frontend store state for drawn cards in a game */
export interface DeckState {
  decks: Deck[];
  currentDeckIndex: number;
}
