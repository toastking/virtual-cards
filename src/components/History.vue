<template>
  <ol class="history-list list">
    <li
      class="list-item has-text-left"
      v-for="entry of historyWithPlayerInfo"
      :key="entry.card"
    >
      <span class="has-text-info">{{ entry.player.name }} </span> drew
      <span class="history-card-info">{{ cardToEmoji(entry.card) }}</span>
      <span class="has-text-grey-light">{{
        prettyPrintTime(entry.timestamp)
      }}</span>
    </li>
  </ol>
</template>

<script lang="ts">
import { State } from '@/store';
import { parse } from 'playing-card-signature';
import Vue from 'vue';
import { mapGetters, mapState, mapActions } from 'vuex';
import { firestore } from 'firebase';
import moment from 'moment';

export default Vue.extend({
  computed: {
    ...mapGetters(['historyWithPlayerInfo']),
  },
  methods: {
    cardToEmoji(card: string) {
      const parsed = parse(card);
      const emojiMapping = { s: '♠️', h: '♥️', c: '♣️', d: '♦️' };
      return `${parsed.rank}${emojiMapping[parsed.suit]}`;
    },
    prettyPrintTime(timestamp: firestore.Timestamp) {
      return moment(timestamp.toDate()).fromNow();
    },
  },
});
</script>

<style scoped>
.history-list {
  height: 25em;
  overflow-y: scroll;
  -webkit-overflow-scrolling: touch;
}

/* Set the font to system font to display emoji properly*/
.history-card-info {
  font-family: sans-serif;
}
</style>
