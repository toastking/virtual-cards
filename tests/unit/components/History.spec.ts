import History from '@/components/History.vue';
import { Avatar, HistoryEntry, Player } from '@/store/state';
import { mount } from '@vue/test-utils';
import { firestore } from 'firebase';
import MockDate from 'mockdate';
import { Store } from 'vuex-mock-store';

describe('History', () => {
  const histories: Array<HistoryEntry & { player: Player }> = [];
  const store = new Store({ getters: { historyWithPlayerInfo: histories } });
  // add other mocks here so they are accessible in every component
  const mocks = {
    $store: store,
  };

  afterEach(() => {
    store.reset();
  });

  describe('snapshots', () => {
    beforeEach(() => {
      MockDate.set(Date.UTC(2020, 5, 17, 16, 30));
    });

    test('empty list', () => {
      store.getters.historyWithPlayerInfo = [];
      const wrapper = mount(History, { mocks });
      expect(wrapper.element).toMatchSnapshot();
    });

    test('a few history elements', () => {
      const mockHistory: Array<HistoryEntry & { player: Player }> = [
        {
          card: '4s',
          player: { avatar: Avatar.ALPACA, name: 'mort' },
          playerId: 'id',
          timestamp: firestore.Timestamp.fromDate(
            new Date(Date.UTC(2020, 5, 17, 18, 29))
          ),
        },
        {
          card: '5h',
          player: { avatar: Avatar.ALPACA, name: 'stove' },
          playerId: 'id',
          timestamp: firestore.Timestamp.fromDate(
            new Date(Date.UTC(2020, 5, 17, 18, 26))
          ),
        },
      ];
      store.getters.historyWithPlayerInfo = mockHistory;

      const wrapper = mount(History, { mocks });
      expect(wrapper.element).toMatchSnapshot();
    });
  });
});
