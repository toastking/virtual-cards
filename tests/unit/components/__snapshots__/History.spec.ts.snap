// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History snapshots a few history elements 1`] = `
<span
  class="history-list list"
  enter-active-class="animate__animated animate__fadeIn"
  leave-active-class="animate__animated animate__fadeOut"
  name="list"
  tag="ol"
>
  <li
    class="list-item has-text-left"
  >
    <span
      class="has-text-info"
    >
      mort 
    </span>
     drew
    
    <span
      class="history-card-info"
    >
      4♠️
    </span>
     
    <span
      class="has-text-grey-light"
    >
      in 2 hours
    </span>
  </li>
  <li
    class="list-item has-text-left"
  >
    <span
      class="has-text-info"
    >
      stove 
    </span>
     drew
    
    <span
      class="history-card-info"
    >
      5♥️
    </span>
     
    <span
      class="has-text-grey-light"
    >
      in 2 hours
    </span>
  </li>
</span>
`;

exports[`History snapshots empty list 1`] = `
<span
  class="history-list list"
  enter-active-class="animate__animated animate__fadeIn"
  leave-active-class="animate__animated animate__fadeOut"
  name="list"
  tag="ol"
/>
`;
